Fix broken gossip acceptance test

spencerkimball opened the pull request:
Restored ClientDisconnectRedundant unittest (it got lost in the shuffle).
Tweaked the usage of Gossip.SetNodeID and Gossip.SetNodeDescriptor
such that Gossip.SetNodeID MUST be called and that SetNodeDescriptor()
isn't sufficient. Previously, either would set the NodeID, which seems
wrong--there should be one way only to do it.

Changed the bootstrapper code to pause between bootstrapping attempts
to prevent busy bootstrapping if the sentinel node simply isn't
available and there's a small gossip network.

Changed Store.maybeGossipFirstRange to include a retry loop to continue
trying with a relatively tight exponential backoff (max backoff is 1s)
in the event that the leader lease is inconclusive. We don't want to be
bound by the 60s cluster ID gossip interval. This is what was tripping
up the gossip peerings test.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3296)

<!-- Reviewable:end -->

spencerkimball committed:
Fix broken gossip acceptance test

Restored ClientDisconnectRedundant unittest (it got lost in the shuffle).
Tweaked the usage of Gossip.SetNodeID and Gossip.SetNodeDescriptor
such that Gossip.SetNodeID MUST be called and that SetNodeDescriptor()
isn't sufficient. Previously, either would set the NodeID, which seems
wrong--there should be one way only to do it.

Changed the bootstrapper code to pause between bootstrapping attempts
to prevent busy bootstrapping if the sentinel node simply isn't
available and there's a small gossip network.

Changed Store.maybeGossipFirstRange to include a retry loop to continue
trying with a relatively tight exponential backoff (max backoff is 1s)
in the event that the leader lease is inconclusive. We don't want to be
bound by the 60s cluster ID gossip interval. This is what was tripping
up the gossip peerings test.

tschottdorf commented on storage/store.go:
```diff
@@ -621,13 +611,31 @@ func (s *Store) startGossip() {
+	}
+	for loop := retry.Start(retryOptions); loop.Next(); {
+		rng := s.LookupReplica(roachpb.RKeyMin, nil)
+		if rng != nil {
```
```go
if rng == nil {
  return nil
}
// rest
```

spencerkimball replied:
Done.

tschottdorf commented on storage/store.go:
```diff
@@ -621,13 +611,31 @@ func (s *Store) startGossip() {
+		rng := s.LookupReplica(roachpb.RKeyMin, nil)
+		if rng != nil {
+			err := rng.maybeGossipFirstRange()
+			if nlErr, ok := err.(*roachpb.NotLeaderError); !ok || nlErr.Leader != nil {
```
seems that with your chance you're never going to get a `nil` `nlErr.Leader` here.

spencerkimball replied:
It's true, but I still wanted to preserve that piece of logic against potential future changes upstream of this call site.

tamird commented:
Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[gossip/client_test.go, line 229 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn-VX7-ZmVmytMOa6:-K4Zn-VX7-ZmVmytMOa7:-899601148) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/client_test.go#L229)):*
nit: `SucceedsWithin` while you're here

---

*[gossip/gossip.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn89laFRTiTC3uBrJ:-K4Zn89laFRTiTC3uBrK:136197372) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L180)):*
`g.is.NodeID == nodeID` by the time this line is executed.

Also, seems like this method should return an error; this condition seems pretty bad.

---

*[gossip/gossip.go, line 506 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn_h4Xwij8JPDQTmR:-K4Zn_h4Xwij8JPDQTmS:-1584797971) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L506)):*
doesn't this still busy loop, just without doing any work?

---

*[storage/replica_test.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4ZoIej90zOT4LG8QDk:-K4ZoIej90zOT4LG8QDl:866407603) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/replica_test.go#L112)):*
why this value?

---

*[storage/store.go, line 624 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4ZoWO5JsHessiTN68Q:-K4ZoWO5JsHessiTN68R:1425259994) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L624)):*
are these comments really helpful?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I saw gossip leak goroutines when it was connecting in a busy-loop (can't be more precise unfortunately, but after recovery the nodes would stick at a couple hundred goroutines of, I think, `gossip.(*server).Gossip()`). Any idea what that is? I'm assuming some fallout of the busy loop, but I didn't track it down further.

---

Reviewed 14 of 16 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 506 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn_h4Xwij8JPDQTmR:-K4Zpg73JhIGId1m9xGF:-1279169979) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L506)):*
there's a `<-g.stalled` below in the omitted part.

---

*[storage/store.go, line 624 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4ZoWO5JsHessiTN68Q:-K4ZqkVlMN1btmP86qFy:1790771722) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L624)):*
they look like we've been copy pasting them around. +1 for removal.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-:-K4Znz28kU6MHCaadx7D:1140867507)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 506 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn_h4Xwij8JPDQTmR:-K4ZuX1lpnAmx4-DruXE:718044452) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L506)):*
But it will busy loop as often as `g.stalled` is signaled. And conversely, it will bootstrap less than the number of writes to `g.stalled`, If `g.stalled` is signaled just before `bootstrapInterval` it will have to wait until the *next* time `g.stalled` is signaled. I would replace this if with another `select` before the read from `g.stalled`:

    select {
        case <-time.After(bootstrapInterval):
        case <-stopper.ShouldStop():
            return
     }

---

*[storage/store.go, line 629 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zx3wZAFyxrYTRK9ZU:-K4Zx3wZAFyxrYTRK9ZV:-1593216544) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L629)):*
This could accumulate a lot of `LeaderLeaseRequests` in the raft pending queue that will all be replayed when the node connects to the cluster. We need to be able to cancel these commands and remove them if they expire while still in the outgoing queue. I'll file a separate issue.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
OK I tracked this down and fixed it. It's a result of the HTTP/RPC stuff--we were waiting to call close callbacks until after every extant RPC had finished. The problem was that hanging gossip fetches would never finish. Now we call the close callbacks immediately if the client connection is closed so we abort those hanging fetches and free up the RPC go routine.

---

Review status: 11 of 17 files reviewed at latest revision, 7 unresolved discussions.

---

*[gossip/gossip.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn89laFRTiTC3uBrJ:-K4ert-dQIa9WFFZaElN:-357997534) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L180)):*
Good catch. Moved this check above the assignment statement.

---

*[gossip/gossip.go, line 506 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn_h4Xwij8JPDQTmR:-K4iyFLeKMgFxlzv2eL0:-1715642963) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L506)):*
OK, took Ben's suggestion.

---

*[storage/replica_test.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4ZoIej90zOT4LG8QDk:-K4c_-wiVx9s1tBCA8be:1438484861) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/replica_test.go#L112)):*
Changed it to 1. It was an artifact of an earlier more restrictive set of checks when setting node ID and node descriptor.

---

*[storage/store.go, line 624 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4ZoWO5JsHessiTN68Q:-K4c_nSUwzMXVIexVjJD:1740281843) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L624)):*
Removed.

---

*[storage/store.go, line 629 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zx3wZAFyxrYTRK9ZU:-K4c_b8sXdNsVcMO7JP9:1651214454) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L629)):*
Isn't that what the context deadline is supposed to be doing already? Does that just not work yet? This is a pretty fringe case, so probably nothing to worry about for the time being.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-:-K4erPzmkHB6gGV6gg6I:-934333233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 16 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[gossip/server.go, line 148 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jp9sW2B0zNm64pSpf:-K4jp9sW2B0zNm64pSpg:-872785651) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/gossip/server.go#L148)):*
is this the one that fixes the goroutine leak?

---

*[gossip/server.go, line 225 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jpG3OY9DvfhxE34Co:-K4jpG3OY9DvfhxE34Cp:785942063) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/gossip/server.go#L225)):*
along with this?

---

*[rpc/server.go, line 205 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jpgrOPP52JluBcJjN:-K4jpgrOPP52JluBcJjO:-1253695864) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/rpc/server.go#L205)):*
doesn't look like this is read from anywhere.

---

*[storage/store.go, line 629 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zx3wZAFyxrYTRK9ZU:-K4jqihvU1GwBWT4zTTS:-1342397557) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L629)):*
we're not using the deadline yet, but issue #3299 which @bdarnell created suggests changing that.

---

*[storage/store.go, line 617 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jqWtIwJbVg7aR_md7:-K4jqWtIwJbVg7aR_md8:1373626738) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/storage/store.go#L617)):*
indicates that

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[gossip/gossip.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn89laFRTiTC3uBrJ:-K4sF6mcxAsRK4xrnzfq:289822682) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L180)):*
It doesn't seem like this should be a `log.Errorf` - probably a panic unless you want to return an error to every call site.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-:-K4sGSEwwtfzndv2I_c9:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Add 'Fixes #3258'.

spencerkimball commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zn89laFRTiTC3uBrJ:-K4xY5YucZ2ZvnM7y1Wj:-1503336344) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/gossip/gossip.go#L180)):*
I added a TODO. There are some complicated issues with the unittests invoking this with different node IDs during same run.

---

*[gossip/server.go, line 148 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jp9sW2B0zNm64pSpf:-K4xX77HPLTQYrf_8Gtn:230558146) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/gossip/server.go#L148)):*
A combination of broadcasting on read when the connection is closed and the changes to rpc/server.go to invoke the close callbacks.

---

*[gossip/server.go, line 225 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jpG3OY9DvfhxE34Co:-K4xXMkJ87MtBFrj9Cmd:160394143) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/gossip/server.go#L225)):*
Yep

---

*[rpc/server.go, line 205 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4jpgrOPP52JluBcJjN:-K4xXZsrSoXs6sO-YiaV:-184265899) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b157078fef792e17a9d0c7062e7267ca2f102a/rpc/server.go#L205)):*
Good point. No longer necessary & removed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[rpc/server.go, line 155 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4xbXmoytIDeJzPoARg:-K4xbXmoytIDeJzPoARh:1342685449) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb0d596fb353df5df1e75d2b641b2deb13f8b1bb/rpc/server.go#L155)):*
what's the point of this nil check?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[rpc/server.go, line 155 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4xbXmoytIDeJzPoARg:-K4xeMfofr-vUlas96pD:1740281843) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb0d596fb353df5df1e75d2b641b2deb13f8b1bb/rpc/server.go#L155)):*
Removed.

---

*[storage/store.go, line 629 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296#-K4Zx3wZAFyxrYTRK9ZU:-K4xeV0KrdOvQTnwHXLz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc08068d8cbe2b55630eb5ad0032c597959c9b4/storage/store.go#L629)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3296)*
<!-- Sent from Reviewable.io -->

mberhault commented:
I'm guessing that this is from this change. I'm now seeing a lot more
gossip-related log spam (eg: hundreds of MB of logs for a 1h run).

One a given node, I get the following (a lot):
I1208 08:05:25.360451 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.28.253:26257
I1208 08:05:25.362736 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.364996 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.28.253:26257
I1208 08:05:25.368646 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.373279 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.28.253:26257
I1208 08:05:25.375338 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.377201 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.383760 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.28.253:26257
I1208 08:05:25.386799 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.28.253:26257
I1208 08:05:25.391384 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.393681 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.395927 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.398422 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.400583 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.402149 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.28.253:26257
I1208 08:05:25.406992 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257
I1208 08:05:25.409657 1514 gossip/server.go:112  received 14 (0 fresh)
info(s) from 172.31.29.253:26257

Another pattern is the following (different node):
I1208 09:05:43.204994 1522 gossip/gossip.go:688  starting client to
172.31.31.97:26257
I1208 09:05:43.211576 1522 gossip/client.go:151  sent 14 info(s) to
172.31.31.97:26257
I1208 09:05:43.212230 1522 gossip/client.go:151  sent 0 info(s) to
172.31.31.97:26257
I1208 09:05:43.212758 1522 gossip/client.go:75  closing client to
172.31.31.97:26257: gossip/client.go:163: received forward from
172.31.31.97:26257 to ip-172-31-29-249:26257
I1208 09:05:43.212783 1522 gossip/gossip.go:688  starting client to
172.31.29.249:26257
I1208 09:05:43.218409 1522 gossip/client.go:151  sent 14 info(s) to
172.31.29.249:26257
I1208 09:05:43.219046 1522 gossip/client.go:151  sent 0 info(s) to
172.31.29.249:26257
I1208 09:05:43.219536 1522 gossip/client.go:75  closing client to
172.31.29.249:26257: gossip/client.go:163: received forward from
172.31.29.249:26257 to ip-172-31-31-97:26257
I1208 09:05:43.219560 1522 gossip/gossip.go:688  starting client to
172.31.31.97:26257
I1208 09:05:43.222726 1522 gossip/client.go:151  sent 0 info(s) to
172.31.31.97:26257
I1208 09:05:43.223250 1522 gossip/client.go:75  closing client to
172.31.31.97:26257: gossip/client.go:163: received forward from
172.31.31.97:26257 to ip-172-31-31-97:26257

In both cases, we seem to be on a 5ms loop or so.

On Mon, Dec 7, 2015 at 5:04 PM, Spencer Kimball <notifications@github.com>
wrote:

> Merged #3296 <https://github.com/cockroachdb/cockroach/pull/3296>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3296#event-484956583>.
>
